src/model.rs

```diff
@@ -206,62 +223,218 @@ impl Analysis<Mdl> for TensorAnalysis {
-            //    let t_b = x(b).meta;
-            //    let t_c = x(c).meta;
-
-            //    unsafe { // very unsafe sketchy
-            //      let cat = g.concat(1, 2, vec![t_b, t_c].as_ptr());
-            //      Self::Data {cost : (*(*cat).op.ptr).runtime, meta : cat}
-            //    }
-            //  },
+
+            Mdl::Weight([name]) => {
+                // Check types
+                assert!(x(name).dtype == DataKind::Name);
+
+                // Get shape
+                let mut split = x(name).name.split("@");
+                let name_vec: Vec<&str> = split.collect();
+                assert!(name_vec.len() == 2);
+                let mut split_dims = name_vec[1].split("_");
+                let dim_str_vec: Vec<&str> = split_dims.collect();
+
+                // Create tensorhandle and get metadata
+                unsafe {
+                    let mut dims: Vec<i32> = dim_str_vec
+                        .iter()
+                        .map(|x| x.parse::<i32>().unwrap())
+                        .collect();
+                    let ndim = dims.len();
+                    assert!(ndim <= 4);
+                    dims.shrink_to_fit();
+                    assert!(dims.len() == dims.capacity());
+
+                    // Create data for weight
+                    let mut num_entries = 1;
+                    for d in &dims {
+                        num_entries = num_entries * d;
+                    }
+                    let mut weight_data: Vec<f32> =
+                        Vec::with_capacity(num_entries.try_into().unwrap());
+                    for _ in 0..weight_data.capacity() {
+                        weight_data.push(rand::random());
+                    }
+                    assert!(weight_data.len() == weight_data.capacity());
+
+                    let ptr = dims.as_mut_ptr();
+                    std::mem::forget(dims);
+                    let data_ptr = weight_data.as_mut_ptr();
+                    std::mem::forget(weight_data);
+
+                    let res = g.new_weight(ndim.try_into().unwrap(), ptr, data_ptr);
+                    Self::Data {
+                        dtype: DataKind::Tnsr,
+                        val: 0,
+                        name: String::new(),
+                        meta: res,
+                    }
+                }
+            }
+
+            Mdl::Concat([axis, ndim, a, b]) => {
+                // Check types
+                assert!(x(axis).dtype == DataKind::Scalar);
+                assert!(x(ndim).dtype == DataKind::Scalar);
+                assert!(x(a).dtype == DataKind::Tnsr);
+                assert!(x(b).dtype == DataKind::Tnsr);
+
+                // Get arguments
+                let t_a = x(a).meta;
+                let t_b = x(b).meta;
+                let axis_val = x(axis).val;
+
+                // Create tensorhandle and get metadata
+                unsafe {
+                    let res = g.concat(axis_val, 2, vec![t_a, t_b].as_ptr());
```

Comment:
* Create a variable to hold `vec![t_a, t_b]`. Otherwise, it might be dealloc before calling concat. Note that this is only needed because you are dealing with pointers. In general, Rust will infer the correct lifetime.
* Since the size is static, you do not need to allocate a vector. Use an array instead (allocated on the stack).
```
let t = [t_a, t_b];
let res = g.concat(axis_val, 2, t.as_ptr());
```

Reply:
Done